lib/runtime/src/ops/transpose.cc

```diff
@@ -50,369 +52,549 @@ TransposeParams Transpose::get_params() const {
+    assert(out_domain.hi()[i] == in_domain.hi()[m->perm[i]]);
+    assert(out_domain.lo()[i] == in_domain.lo()[m->perm[i]]);
+  }
+
+  return profiling(forward_kernel,
+                   profiling,
+                   "[Transpose] Forward_time = %.2lf [ms]",
+                   &per_device_state,
+                   input.get_float_ptr(),
+                   output.get_float_ptr(),
+                   in_domain,
+                   out_domain);
 }
 
-void Transpose::forward(FFModel const &ff) {
-  ArgumentMap argmap;
-  Context ctx = ff.config.lg_ctx;
-  Runtime *runtime = ff.config.lg_hlr;
-  set_argumentmap_for_forward(ff, argmap);
-  IndexLauncher launcher(TRANSPOSE_FWD_TASK_ID,
-                         parallel_is,
-                         TaskArgument(NULL, false),
-                         argmap,
-                         Predicate::TRUE_PRED,
-                         false /*must*/,
-                         0 /*mapper_id*/,
-                         outputs[0]->machine_view.hash());
-  launcher.add_region_requirement(RegionRequirement(inputs[0]->part,
-                                                    0 /*projection id*/,
-                                                    READ_ONLY,
-                                                    EXCLUSIVE,
-                                                    inputs[0]->region));
-  launcher.add_field(0, FID_DATA);
-  launcher.add_region_requirement(RegionRequirement(outputs[0]->part,
-                                                    0 /*projection id*/,
-                                                    WRITE_ONLY,
-                                                    EXCLUSIVE,
-                                                    outputs[0]->region));
-  launcher.add_field(1, FID_DATA);
-  runtime->execute_index_space(ctx, launcher);
+static void forward_task(Task const *task,
+                         std::vector<PhysicalRegion> const &regions,
+                         Context ctx,
+                         Runtime *runtime) {
+  TaskArgumentAccessor acc(task, regions, ctx, runtime);
+  forward_task_impl(acc);
 }
 
-void Transpose::forward_task(Task const *task,
-                             std::vector<PhysicalRegion> const &regions,
-                             Context ctx,
-                             Runtime *runtime) {
-  assert(regions.size() == 2);
-  assert(task->regions.size() == 2);
-  // const Transpose* transpose = (const Transpose*) task->args;
-  TransposeMeta const *m = *((TransposeMeta **)task->local_args);
+static optional<float> backward_task_impl(TaskArgumentAccessor const &acc) {
+  ProfilingSettings profiling = acc.get_argument<ProfilingSettings>(PROFILING);
+  auto per_device_state =
+      acc.get_per_device_state<TransposePerDeviceState>(PER_DEVICE_STATE);
+
+  auto input_grad = acc.get_tensor_grad<Permissions::RO>(INPUT);
+  auto output_grad = acc.get_tensor_grad<Permissions::WO>(OUTPUT);
+
+  // copy from the old code
+  Task *task = acc.task;
```

Comment:
this may have problem

```
  Task *task = acc.task;
  Context ctx = acc.ctx;
  Runtime *runtime = acc.runtime;
  Domain in_domain = runtime->get_index_space_domain(
      ctx, task->regions[0].region.get_index_space());
  Domain out_domain = runtime->get_index_space_domain(
      ctx, task->regions[1].region.get_index_space());

  for (int i = 0; i < m->num_dim; i++) {
    assert(out_domain.hi()[i] == in_domain.hi()[m->perm[i]]);
    assert(out_domain.lo()[i] == in_domain.lo()[m->perm[i]]);
  }
```